### ISSUE TITLE
WARNING sqlfluff.linter:linter.py:525 Fixes for ST07 not applied, as it would result in an unparsable file

### ISSUE THREAD

**Issue by Pymancer** [sqlfluff/sqlfluff#6043]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

sqlfluf.fix asked to report a bug in the warning

### Expected Behaviour

No warnings

### Observed Behaviour

 Warning issued by sqlfluff.fix:
```sh
Fixes for ST07 not applied, as it would result in an unparsable file. Please report this as a bug with a minimal query which demonstrates this warning.
One fix for ST07 not applied, it would re-cause the same error.
```

### How to reproduce

sqlfluff.fix('select "tar"."name" from "bar" full join "tar" using ("id","code") full join "car" using ("id")', dialect='postgres')

### Dialect

postgres

### Version

3.1.0

### Configuration

No custom configuration

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by Pymancer** [comment]
Another similar warning:
```sh
Fixes for RF03 not applied, as it would result in an unparsable file. Please report this as a bug with a minimal query which demonstrates this warning.
Fixes for RF03 not applied, as it would result in an unparsable file. Please report this as a bug with a minimal query which demonstrates this warning.
One fix for RF03 not applied, it would re-cause the same error.
```
on
```sh
sqlfluff.fix('select "t"."id","t"."code" from (select 1 as "id",\'foo\' as "code" from "bar") as "t" where "code" = \'foo\' order by "code"', dialect='postgres')
```

**Comment by peterbud** [comment]
With the latest `sqlfluff` the above statement is running properly through fix:

```sql
select "tar"."name" from "bar" full join "tar" using ("id","code") full join "car" using ("id")
```

```bash
sqlfluff fix test.sql -d postgres
==== finding fixable violations ====
== [test.sql] FAIL                                                                                                                                                                                              
L:   1 | P:   1 | LT05 | Line is too long (96 > 80). [layout.long_lines]
L:   1 | P:   8 | RF06 | Unnecessary quoted identifier "tar".
                       | [references.quoting]
L:   1 | P:  14 | RF06 | Unnecessary quoted identifier "name".
                       | [references.quoting]
L:   1 | P:  26 | RF06 | Unnecessary quoted identifier "bar".
                       | [references.quoting]
L:   1 | P:  42 | RF06 | Unnecessary quoted identifier "tar".
                       | [references.quoting]
L:   1 | P:  48 | ST07 | Found USING statement. Expected only ON statements.
                       | [structure.using]
L:   1 | P:  55 | RF06 | Unnecessary quoted identifier "id".
                       | [references.quoting]
L:   1 | P:  60 | LT01 | Expected single whitespace between comma ',' and quoted
                       | identifier. [layout.spacing]
L:   1 | P:  60 | RF06 | Unnecessary quoted identifier "code".
                       | [references.quoting]
L:   1 | P:  78 | RF06 | Unnecessary quoted identifier "car".
                       | [references.quoting]
L:   1 | P:  91 | RF06 | Unnecessary quoted identifier "id".
                       | [references.quoting]
L:   1 | P:  96 | LT12 | Files must end with a single trailing newline.
                       | [layout.end_of_file]
== [test.sql] FIXED
12 fixable linting violations found
  [3 unfixable linting violations found]
```

Result:
```sql
select tar.name
from bar
full join tar on bar.id = tar.id and bar.code = tar.code
full join car using (id)
```
